### DISCUSSION TITLE
Spec V2 idea: support custom assembler / disassembler

### DISCUSSION THREAD

**Discussion by sluongng** [git-lfs/git-lfs#5132] [Ideas]
After tracing down threads such as 
- https://github.com/git-lfs/git-lfs/issues/355
- https://github.com/git-lfs/git-lfs/issues/4819

It might be time to think about what a spec V2 could look like.

I am especially interested in how we could divide 1 big file into multiple smaller chunks to accomplish 2 goals:

1. Faster download through parallelism. You could fetch multiple chunks in parallel to maximize your download network throughputs. 
2. Resumable download. Smaller chunks should also help download be resumable when the network condition worsens. You would only need to redownload the smaller chunk instead of having to redownload everything.
3. Support packed storage. By dividing large files into smaller chunks, we could re-use sets of chunks that are identical between versions of the files, or between different copies of the same file. This helps with creating an efficient `pack` of blobs down the line.

For this to happen, I think we would want to introduce a Spec V2 where optional binaries could be specified as `assembler` and `disassembler` for a large file.

- Disassembler: ingest 1 file and divide it into smaller chunks using custom logic. Chunks are then written into Git-LFS CAS and a pointer file with spec v2 is written with a list of chunks.
- Assembler: put together a file from the chunks listed in the pointer file.

Both Disassembler / Assembler binaries could opt to implement their own upload/download implementation so that advanced implementations could take advantage of techniques such as prefetch, prepush, p2p transfer or erasure encoding.

An example of what I have in mind:

```
version https://git-lfs.github.com/spec/v2
oid sha256:4d7a214614ab2935c943f9e0ff69d22eadbb8f32b1258daaa5e2ca24d17e2393
size 12345
assembler: foo-asmblr
disassembler: foo-dasmblr
chunk sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
size 4115
chunk sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
size 4115
chunk sha256:ccccccccccccccccccccccccccccccccccccccccccccccc
size 4115
```

**Comment by sluongng** [comment]
Prior art: `Arc` version control of Yandex could optionally replace a blob object with `BlobRef`, which is a reference to smaller chunks of blobs

https://habr.com/ru/company/yandex/blog/482926/ (use google translate)

**Comment by bk2204** [comment]
Hey,

I think we could in fact implement a multipart upload or download approach, and we have an accepted (but not yet implemented) proposal for that in #4438.  However, I don't think we want to expose this fact in the pointer files, which we rely on being very small (less than 1 KiB), nor do we want to make this arbitrarily configurable to use arbitrary assemblers and disassemblers.

In general, specifying arbitrary programs to assemble and disassemble requires configuration on the part of the user since shipping configuration in the repo that invokes a particular program is a security problem.  Therefore, specifying a custom assembler and disassembler for pointer files means those wouldn't work out of the box and thus things would be a lot less useful for users.  Specifying a generic multipart approach is more flexible and puts the configuration on the server, which the user doesn't have to know about.

Of course, users can already implement a custom upload and download approach if they want.  Our `file:///` protocol handling uses that mechanism itself.
